libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_OABendyRuler.h"
+#include <AC_Fence/AC_Fence.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL &hal;
+
+const int16_t OA_BENDYRULER_BEARING_INC = 5;            // check every 5 degrees around vehicle
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_RATIO = 1.0f; // step2's lookahead length as a ratio of step1's lookahead length
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_MIN = 2.0f;   // step2 checks at least this many meters past step1's location
+const float OA_BENDYRULER_LOOKAHEAD_PAST_DEST = 2.0f;   // lookahead length will be at least this many meters past the destination
+const float OA_BENDYRULER_LOW_SPEED_SQUARED = 0.04f;    // when ground course is below this speed squared, vehicle's heading will be used (0.04 = 0.2m/s)
+
+// run background task to find best path and update avoidance_results
+// returns true and updates origin_new and destination_new if a best path has been found
+bool AP_OABendyRuler::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new)
```

Comment:
ok, point taken... but in Copter and Rover we very commonly use "origin" when referring to the starting point of a journey to a waypoint.  I.e. the start of the segment.  We could perhaps move to "source" instead of "origin" but it's a relatively large renaming exercise that I don't think I'll take on as part of this PR.